### PR TITLE
Fixes predictors in the multitask case

### LINKED ISSUE #4973: `allennlp predict` is broken for multitask models
See https://stackoverflow.com/questions/66156046/allennlp-2-0-using-allennlp-predict-with-multitaskdatasetreader-leads-to-runt.

### PR DESCRIPTION
**Pull request by dirkgr**
Fixes #4973.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -20,6 +20,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added `ModelUsage` to `ModelCard` class.
 - Added a way to specify extra parameters to the predictor in an `allennlp predict` call.
 - Added a way to initialize a `Vocabulary` from transformers models.
+- Added the ability to ise `Predictors` with multitask models.
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply CHANGELOG.md:23]: Maybe you should mention the new class: `MultiTaskPredictor`

File: allennlp/predictors/multitask.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    """
+    Predictor for multitask models.
+
+    Registered as a `Predictor` with name "multitask".
+    """
+
+    _WRONG_READER_ERROR = (
+        "MultitaskPredictor is designed to work with MultiTaskDatasetReader. "
+        + "If you have a different DatasetReader, you have to write your own "
+        + "Predictor, but you can use MultitaskPredictor as a starting point."
```
Comment (epwalsh) [inline allennlp/predictors/multitask.py:21]:
```suggestion
        + "Predictor, but you can use MultiTaskPredictor as a starting point."
```

File: allennlp/predictors/multitask.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+from allennlp.predictors.predictor import Predictor
+
+
+@Predictor.register("multitask")
+class MultiTaskPredictor(Predictor):
+    """
+    Predictor for multitask models.
+
+    Registered as a `Predictor` with name "multitask".
+    """
```
Comment (epwalsh) [inline allennlp/predictors/multitask.py:16]: I think we should also state here that it only works with `MultiTaskDatasetReader` and `MultiTaskModel`

File: allennlp/predictors/multitask.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        + "If you have a different DatasetReader, you have to write your own "
+        + "Predictor, but you can use MultitaskPredictor as a starting point."
+    )
+
+    _WRONG_FIELD_ERROR = (
+        "MultiTaskPredictor expects instances that have a MetadataField "
+        + "with the name 'task', containing the name of the task the instance is for."
+    )
+
+    def __init__(self, model: Model, dataset_reader: DatasetReader) -> None:
```
Comment (epwalsh) [inline allennlp/predictors/multitask.py:29]: Can we specify the type restriction in the type hints too?

```suggestion
    def __init__(self, model: MultiTaskModel, dataset_reader: MultiTaskDatasetReader) -> None:
```

File: allennlp/predictors/multitask.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                Predictor.by_name(predictor_name) if predictor_name is not None else Predictor  # type: ignore
+            )
+            self.predictors[name] = predictor_class(model, dataset_reader.readers[name])
+
+    @overrides
+    def predict_instance(self, instance: Instance) -> JsonDict:
+        from allennlp.data.fields import MetadataField
+        from allennlp.common.util import sanitize
+        from allennlp.data.dataset_readers import MultiTaskDatasetReader
+        from allennlp.common.checks import ConfigurationError
```
Comment (epwalsh) [inline allennlp/predictors/multitask.py:57]: What's the rational for putting these imports here? I think it adds unnecessary runtime overhead.

File: allennlp/predictors/multitask.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        task: str = task_field.metadata
+        if not isinstance(self._dataset_reader, MultiTaskDatasetReader):
+            raise ConfigurationError(self._WRONG_READER_ERROR)
+        self._dataset_reader.readers[task].apply_token_indexers(instance)
+        outputs = self._model.forward_on_instance(instance)
+        return sanitize(outputs)
+
+    @overrides
+    def _json_to_instance(self, json_dict: JsonDict) -> Instance:
+        from allennlp.data.fields import MetadataField
```
Comment (epwalsh) [inline allennlp/predictors/multitask.py:71]: Same question about imports here.

File: allennlp/predictors/multitask.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        instance = predictor._json_to_instance(json_dict)
+        instance.add_field("task", MetadataField(task))
+        return instance
+
+    @overrides
+    def predict_batch_instance(self, instances: List[Instance]) -> List[JsonDict]:
+        import collections
+        from allennlp.data.fields import MetadataField
+        from allennlp.common.checks import ConfigurationError
+        from allennlp.data.dataset_readers import MultiTaskDatasetReader
```
Comment (epwalsh) [inline allennlp/predictors/multitask.py:85]: And here

File: allennlp/modules/backbones/vilbert_backbone.py [hunk 8]
```diff
@@ -215,20 +215,3 @@ def forward(
...
-            "token_ids"
-        ]:  # TODO: do we even have "token_ids" in the output?
-            tokens.append(
-                [
-                    self._vocab.get_token_from_index(token_id.item(), namespace=self._namespace)
-                    for token_id in instance_tokens
-                ]
-            )
-        output_dict["tokens"] = tokens
-        return output_dict
```
Comment (epwalsh) [inline allennlp/modules/backbones/vilbert_backbone.py:234]: What happened here?

**Comment by epwalsh** [comment]
Is it feasible to add a test for `MultiTaskPredictor`? Or at least test it as part of the `MultiTaskModel` tests?

**Comment by dirkgr** [comment]
> Is it feasible to add a test for MultiTaskPredictor?

It's being tested from the models repo. We don't have any multitask examples in the core repo, so we can't test it here.

File: allennlp/modules/backbones/vilbert_backbone.py (hunk 8 above)
Comment (dirkgr) [reply allennlp/modules/backbones/vilbert_backbone.py:234]: That never worked. It never ran. It was probably copy and pasted into this class and then never checked.

File: allennlp/predictors/multitask.py (hunk 5 above)
Comment (dirkgr) [reply allennlp/predictors/multitask.py:57]: We have problems with circular imports in many places. I work around them by using local imports for anything allennlp when I can.

This also helps with startup times.

File: allennlp/predictors/multitask.py (hunk 4 above)
Comment (dirkgr) [reply allennlp/predictors/multitask.py:29]: Turns out we can! Yay!

**Comment by dirkgr** [comment]
@epwalsh I think I addressed all the comments. This is ready for another review.

File: allennlp/predictors/multitask.py (hunk 5 above)
Comment (epwalsh) [reply allennlp/predictors/multitask.py:59]: I *really* don't think this is good practice, especially in a function that is meant to be called repeatedly. And none of these imports should cause circular import errors in this module anyway.

File: CHANGELOG.md [hunk 9]
```diff
@@ -17,6 +17,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Added `ModelUsage` to `ModelCard` class.
 - Added a way to specify extra parameters to the predictor in an `allennlp predict` call.
 - Added a way to initialize a `Vocabulary` from transformers models.
+- Added the ability to use `Predictors` with multitask models.
```
Comment (epwalsh) [inline CHANGELOG.md:20]:
```suggestion
- Added the ability to use `Predictors` with multitask models through the new `MultiTaskPredictor`.
```

**Review by epwalsh — DISMISSED** [review]
LGTM!

File: CHANGELOG.md (hunk 9 above)
Comment (dirkgr) [reply CHANGELOG.md:20]: Oh did I forget to do that? I meant to. Sorry!

**Review by epwalsh — APPROVED** [review]

**Pull request merged**